cpp/src/fil/fil.cu

```diff
@@ -636,6 +637,89 @@ int tree2fil_sparse(std::vector<fil_node_t>& nodes, int root,
+inline void tree_depth_hist(const tl::Tree<T, L>& tree,
+                            std::unordered_map<int, level_entry>& hist) {
+  std::stack<pair_t> stack;  // {tl_id, depth}
+  stack.push({tree_root(tree), 0});
+  while (!stack.empty()) {
+    int node_id, depth;
+    tie(node_id, depth) = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      hist[depth].n_branches++;
+      stack.push({tree.LeftChild(node_id), depth + 1});
+      node_id = tree.RightChild(node_id);
+      depth++;
+    }
+    hist[depth].n_leaves++;
+  }
+}
+
+// implements https://tools.ietf.org/html/draft-eastlake-fnv-17.html
+// algorithm is public domain without export controls and no patents or rights to patent
+auto fowler_noll_vo_fingerprint64(int* begin, int* end) {
+  return std::accumulate(begin, end, 14695981039346656037lu,
+                         [](auto& fingerprint, int& x) {
+                           return (fingerprint * 1099511628211lu) ^ x;
+                         });
+}
+
+template <typename T, typename L>
+std::stringstream depth_hist_and_max(const tl::ModelImpl<T, L>& model) {
+  std::unordered_map<int, level_entry> depth_hist;
+  for (const auto& tree : model.trees) tree_depth_hist(tree, depth_hist);
+  struct hist_entry {
+    int level, n_branches, n_leaves;
+  };
+  std::vector<hist_entry> vec_hist;
+  for (auto p : depth_hist) {
+    vec_hist.emplace_back(
+      hist_entry{p.first, p.second.n_branches, p.second.n_leaves});
+  }
+  std::sort(vec_hist.begin(), vec_hist.end(),
+            [](auto a, auto b) { return a.level < b.level; });
+
+  int min_depth = -1, leaves_times_depth = 0, total_branches = 0,
+      total_leaves = 0;
+  std::stringstream forest_shape;
+  forest_shape << "Depth histogram:\n"
+               << "depth branches leaves   nodes\n";
+  char line[100];
+  for (hist_entry e : vec_hist) {
+    sprintf(line, "%5d %8d %6d %7d\n", e.level, e.n_branches, e.n_leaves,
+            e.n_branches + e.n_leaves);
+    forest_shape << line;
+    if (e.n_leaves && min_depth == -1) min_depth = e.level;
+    leaves_times_depth += e.n_leaves * e.level;
+    total_branches += e.n_branches;
+    total_leaves += e.n_leaves;
+  }
+  int total_nodes = total_branches + total_leaves;
+  sprintf(line, "Total: branches: %d leaves: %d nodes: %d\n", total_branches,
+          total_leaves, total_nodes);
+  forest_shape << line;
+  sprintf(line, "Avg nodes per tree: %.1f\n",
+          total_nodes / (float)vec_hist[0].n_branches);
+  forest_shape << line;
+  sprintf(line, "Leaf depth: min: %d avg %.1f max %lu\n", min_depth,
+          leaves_times_depth / (float)total_leaves, vec_hist.size() - 1);
+  forest_shape << line;
+  // std::hash does not promise to not be identity. Xoring plain numbers which
+  // add up to one another erases information, hence, std::hash is unsuitable here
+  sprintf(line, "Depth histogram fingerprint: %0lx\n",
+          fowler_noll_vo_fingerprint64((int*)&*vec_hist.begin(),
```

Comment:
Please avoid such pointer casting. It's better to create an extra array of `int`.